src/bp-core/compatibility/bp-wp-offload-media-helpers.php

```diff
@@ -0,0 +1,325 @@
+			$get_metadata = wp_get_attachment_metadata( $attachment_id );
+			if ( ! empty( $get_metadata ) && isset( $get_metadata['sizes'] ) && isset( $get_metadata['sizes'][ $size ] ) ) {
+				$attachment_url = wp_get_attachment_image_url( $attachment_id, $size );
+			} else {
+				$attachment_url = wp_get_attachment_image_url( $attachment_id, 'full' );
+			}
+
+			$image_array = @getimagesize( $attachment_url );
+
+			if ( ! $attachment_url || empty( $image_array ) ) {
+
+				add_filter( 'as3cf_get_attached_file_copy_back_to_local', array( $this, 'bb_document_as3cf_get_attached_file_copy_back_to_local' ), PHP_INT_MAX, 4 );
+				add_filter( 'as3cf_upload_acl', array( $this, 'bb_media_private_upload_acl' ), 10, 1 );
+				add_filter( 'as3cf_upload_acl_sizes', array( $this, 'bb_media_private_upload_acl' ), 10, 1 );
+
+				bp_document_generate_document_previews( $attachment_id );
+
+				remove_filter( 'as3cf_upload_acl', array( $this, 'bb_media_private_upload_acl' ), 10, 1 );
+				remove_filter( 'as3cf_upload_acl_sizes', array( $this, 'bb_media_private_upload_acl' ), 10, 1 );
+				remove_filter( 'as3cf_get_attached_file_copy_back_to_local', array( $this, 'bb_document_as3cf_get_attached_file_copy_back_to_local' ), PHP_INT_MAX, 4 );
+
+				if ( ! empty( $get_metadata ) && isset( $get_metadata['sizes'] ) && isset( $get_metadata['sizes'][ $size ] ) ) {
+					$attachment_url = wp_get_attachment_image_url( $attachment_id, $size );
+				} else {
+					$attachment_url = wp_get_attachment_image_url( $attachment_id, 'full' );
+				}
+			}
+		}
+
+		if ( ! $symlink && in_array( $extension, bp_get_document_preview_music_extensions(), true ) ) {
+			if ( ! empty( $attachment_id ) && ! empty( $document_id ) ) {
+				$attachment_url = wp_get_attachment_url( $attachment_id );
+			}
+		}
+
+		return $attachment_url;
+	}
+
+	/**
+	 * Return the offload media plugin attachment url.
+	 *
+	 * @param string $attachment_url attachment url.
+	 * @param int    $media_id       media id.
+	 * @param int    $attachment_id  attachment id.
+	 * @param string $size           size of the media.
+	 * @param bool   $symlink        display symlink or not.
+	 *
+	 * @return false|mixed|string return the original media URL.
+	 *
+	 * @since BuddyBoss 1.7.0
+	 */
+	public function bp_media_offload_get_preview_url( $attachment_url, $media_id, $attachment_id, $size, $symlink ) {
+
+		if ( ! $symlink ) {
+			$media          = new BP_Media( $media_id );
+			$attachment_url = wp_get_attachment_url( $media->attachment_id );
+		}
+
+		return $attachment_url;
+	}
+
+	/**
+	 * Make all the media to private signed URL if someone using the offload media to store in AWS.
+	 *
+	 * @handles `as3cf_upload_acl`
+	 * @handles `as3cf_upload_acl_sizes`
+	 *
+	 * @param string $acl defaults to 'public-read'.
+	 *
+	 * @return string $acl make the media to private with signed url.
+	 *
+	 * @since   BuddyBoss 1.7.0
+	 */
+	public function bb_media_private_upload_acl( $acl ) {
+		$acl = 'private';
```

Comment:
This is problematic as it will only work for S3-API compatible storage providers that allow canned ACLs.

For example, if the site is using Google Cloud Storage (GCS) then you'd need to return "projectPrivate" instead.

I'll offer an alternative solution in my review overview.

Reply:
@ianmjones We are uploading all the media in our own folders called `bb_medias, bb_documents & bb_videos` respectively under the `wp-content/uploads/` directory and we have different privacy options (public, friends, onlyme, members only) for all the uploaded media so we are making those media privately and not accessible via direct link that's why we are serving the media via `symlink` OR `rewrite_rules` based when not using the offload media.

So we need to add all the media uploaded via BuddyBoss Platform that should be in the private (signed URL) we recommending so it will follow the privacy for some time interval based on the user selected as mentioned above.

Are there any other options you are using to make the media private `private` & `projectPrivate` except this?

Reply:
To answer your question, I need to ask you another first ...

If WP Offload Media wasn't in the mix, would those privacy options change whether a URL can be accessed, or just who can discover the URL?

To clarify ... if a photo was set as "friends" only, but a friend shared the photo's URL with someone that doesn't even have an account, would they still see the image?

Reply:
@ianmjones If the server is using the `rewrite_rules` and if the user shares the URL with other members then it will not show the preview. But if the server is using the `symlink` then it will show the preview of the file but we are still enhancing the `symlink` to change the URL every 24 hours to expiry so if they share the URL then it will not going to work after a certain time.

Reply:
Thanks @chetansatasiya.

We've already seen a support request from a shared customer that has been affected by the way BuddyBoss already tries to set offloads as private and gets it wrong for the storage provider they are using.

Unfortunately at the moment there is no supported way for a third party plugin like BuddyBoss to tell WP Offload Media that a specific Media Library item should be offloaded as private, and we're not sure there really should be, as generally only WP Offload Media should even know that such a concept exists.

For plugins like EDD and WooCommerce that have similar requirements to BuddyBoss of needing to restrict access to files, they use similar mechanisms of not showing file URLs to unauthorized site visitors. It's then WP Offload Media's responsibility to change the URL that is shown to authorized site visitors to be for the offloaded file as appropriate, and optionally a signed URL that expires within a few seconds. That's why WP Offload Media has built in integration for EDD and WooCommerce so that it can set the files that those plugins are using as Product Files as private in the bucket.

There are potential complications that arise from setting objects as private, and very real financial implications if you set a frequently accessed object as private without also correctly setting up signed CloudFront URLs etc.

As such, we believe BuddyBoss should not be setting the ACL at all. Instead, WP Offload Media should potentially set the files it offloads as private when it recognizes that the source is protected. We'll add it to our todo list to investigate adding setting offloaded BuddyBoss media items as private in a future update to the BuddyBoss integration we already have planned. We'll likely be in touch to check the best way to ask BuddyBoss whether an item needs restricted access.

In the mean time, WP Offload Media users already have access to filters they can use to set offloaded items as private in the bucket when offloaded, with full control of which files should or should not incur the overhead, and properly set the correct ACL for their chosen storage provider.